### PR TITLE
Data robustness during restart

### PR DESCRIPTION
**Pull request by Leo-Besancon**
* [ ] document all added functions
* [ ] try in sandbox /simulation/labnet
* [ ] unit tests on the added/changed features
  * [ ] make tests compile
  * [ ] make tests pass 
* [ ] add logs allowing easy debugging in case the changes caused problems
* [ ] if the API has changed, update the API specification

### CODE REVIEW

**Comment by Leo-Besancon** [comment]
@aoudiamoncef, if you think it's relevant to add a context info that we are in downtime on the API response of get_stakers, please guide me on how to do so ;)

**Comment by aoudiamoncef** [comment]
@Leo-Besanconwe have two ways to bring it to Massa, using a field in the response or a status codes. 

JsonRPC https://www.jsonrpc.org/specification
gRPC https://grpc.github.io/grpc/core/md_doc_statuscodes.html

I'm voting for a status field in the response which is explicit

File: massa-final-state/src/final_state.rs [hunk 1]
```diff
@@ -287,18 +287,20 @@ impl FinalState {
...
     ) -> Result<(), FinalStateError> {
-        let latest_snapshot_cycle =
-            self.pos_state
-                .cycle_history_cache
-                .pop_back()
-                .ok_or(FinalStateError::SnapshotError(String::from(
-                    "Invalid cycle_history",
-                )))?;
+        let latest_snapshot_cycle = self.pos_state.cycle_history_cache.back().cloned().ok_or(
+            FinalStateError::SnapshotError(String::from("Invalid cycle_history")),
```
Comment (aoudiamoncef) [inline massa-final-state/src/final_state.rs:291]: Don't hesitate to add more détails in error messages, this is very helpful

File: massa-pos-exports/src/pos_final_state.rs [hunk 2]
```diff
@@ -711,6 +711,10 @@ impl PoSFinalState {
     pub fn get_all_roll_counts(&self, cycle: u64) -> BTreeMap<Address, u64> {
         let db = self.db.read();
 
+        if self.get_cycle_index(cycle).is_none() {
+            panic!("Cycle {} not in history", cycle)
```
Comment (aoudiamoncef) [inline massa-pos-exports/src/pos_final_state.rs:715]: Is It mandatory to kill the node ?

**Review by aoudiamoncef — APPROVED** [review]

File: massa-pos-exports/src/pos_final_state.rs (hunk 2 above)
Comment (Leo-Besancon) [reply massa-pos-exports/src/pos_final_state.rs:715]: Not really mandatory, but we should never query this info on a cycle that is not on disk. If it's the case, I would not trust the state our node is in, so we should panic to be safe and avoid unexpected consequences of that IMO.

**Comment by aoudiamoncef** [comment]
@Leo-Besancon please don't forget to propagate the changes to `GetLargestStaker` in gRPC API

**Review by aoudiamoncef — APPROVED** [review]
LGTM 👍🏻

**Pull request merged**